finality-aleph/src/finalization.rs

```diff
@@ -77,194 +80,133 @@ where
-        let mut client = Arc::new(TestClientBuilder::new().build());
-
-        let n = 5;
-        let blocks = create_chain(&mut client, n as u64);
-
-        for i in 0..=n {
-            for j in 0..i {
-                let maybe_data = should_finalize(
-                    blocks[i],
-                    AlephData::new(blocks[j], j as u64),
-                    client.as_ref(),
-                    100u64,
-                );
-                assert!(maybe_data.is_none());
+            // At this point we know that data validation went through, which means that the blocks are available
+            // and the whole `new_data` is honest and correct. Note that there might be cases in which some
+            // other node actually failed data validation for `new_data`. This could have happened in two cases:
+            // 1) Data availability let new_data through even though the `proposal` `new_data.branch.last()` was not available
+            // because it could infer that the `proposal` couldn't be finalized anyway (some conflicting block was finalized).
+            // 2) The whole `new_data.branch` was available and data availability let it through, but somehow between that moment and
+            // the moment we obtained `new_data` here, some blocks from the branch were pruned (this, most likely, can happen only with pruning
+            // is enabled).
+            //
+            // Both these cases are handled properly because if data validation does not fail, then the `last_finalized` is guaranteed
+            // to be a block conflicting with `new_data.branch.last()` and hence we will output `None` anyway.
+
+            let proposed_block = proposal.top_block();
+
+            // Early skip for optimization
+            if proposed_block.hash == last_finalized.hash
+                || proposed_block.num <= last_finalized.num
+            {
+                return None;
             }
-        }
 
-        let extra_children: Vec<_> = blocks
-            .iter()
-            .map(|hash| {
-                let mut builder = client
-                    .new_block_at(&BlockId::Hash(*hash), Default::default(), false)
-                    .unwrap();
-                // Add a dummy extrinsic to make the block distinct from the one on chain
-                builder
-                    .push(Extrinsic::AuthoritiesChange(Vec::new()))
-                    .unwrap();
-                let block = builder.build().unwrap().block;
-                let hash = block.header.hash();
-                futures::executor::block_on(client.import(BlockOrigin::Own, block)).unwrap();
-                hash
-            })
-            .collect();
+            // Iterate ancestors of `proposed_hash` until reaching a block with number <= last_finalized_number
+            // in order to check if proposed_block.hash is an ancestor of last_finalized
+            let mut hash = proposed_block.hash;
+            loop {
```

Comment:
Am I right that the bound `MAX_DATA_BRANCH_LEN` does not prevent us from many (session length) iterations here?

Reply:
This has been changed and now is much improved. Take a look.